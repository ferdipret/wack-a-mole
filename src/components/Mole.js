import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import media from "./media"

/**
 * The margin calculation is taking the combined width of the row's
 * elements, subtracting that from the total viewport width to give
 * the remainder, and deviding the remainder into 8(left and right margins
 * for each of the 4 elements)
 *
 * The reason we do it this way, is so we have a fixed width for
 * the images, since they will scale differently, we want to have more control
 * over them
 */
export const MoleContainer = styled.div`
  border-radius: 50%;
  cursor: crosshair;
  display: flex;
  height: 75px;
  margin: 20px calc((100% - (4*75px)) / 8);
  position: relative;
  width: 75px;

  &:active {
    background-color: ${props => props.success && "hsl(120, 50%, 60%)"};
  }

  ${media.tablet`
    height: 150px;
    margin: 20px calc((100% - (4*150px)) / 8);
    width: 150px;
  `};
`

export const MoleWrapper = styled.div`
  display: block
  height: 80%;
  left: 0;
  overflow: hidden;
  position: absolute;
  top: 0;
  width: 100%;
`

export const BombWrapper = styled.div`
  display: block
  height: 80%;
  left: 0;
  overflow: hidden;
  position: absolute;
  top: 0;
  width: 100%;
`

export const BombImg = styled.div`
  background: url("http://clipartist.net/openclipart.org/2013/Mo/AhNinn/bomb-555px.png") no-repeat center;
  background-size: contain;
  height: 100%;
  left: 0;
  position: absolute;
  top: ${({active}) => active ? 0 : "80px"};
  width: 100%;

  transition: all 0.2s ease-in;
`

export const MoleImg = styled.div`
  background: url("http://katherinekato.com/misc/mole.svg") no-repeat center;
  background-size: contain;
  height: 100%;
  left: 0;
  position: absolute;
  top: ${({active}) => active ? 0 : "80px"};
  width: 100%;

  transition: all 0.2s ease-in;
`

export const DirtImg = styled.div`
  background: url("http://katherinekato.com/misc/dirt.svg")
    no-repeat center bottom 5px;
  background-size: contain;
  height: 100%;
  left: 0;
  position: absolute;
  width: 100%;
`

function Mole ({ active, index, onMoleClick, bomb }) {
  const onClick = () => {
    onMoleClick(index)
  }

  return (
    <MoleContainer
      className="mole-container"
      onClick={onClick}
      success={active}>
      <BombWrapper>
        <BombImg className="bomb-image" active={bomb}/>
      </BombWrapper>
      <MoleWrapper>
        <MoleImg className="mole-image" active={active}/>
      </MoleWrapper>
      <DirtImg />
    </MoleContainer>
  )
}

Mole.propTypes = {
  active: PropTypes.bool,
  bomb: PropTypes.bool,
  index: PropTypes.number.isRequired,
  onMoleClick: PropTypes.func.isRequired,
}

Mole.defaultProps = {
  active: false,
}

export default Mole
