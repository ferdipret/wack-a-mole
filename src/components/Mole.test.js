import React from "react"
import { shallow } from "enzyme"
import Mole from "./Mole"
import "jest-styled-components"

describe("Mole", () => {
  const onMoleClick = jest.fn()
  const props = {
    index: 1,
    onMoleClick,
  }

  const wrapper = shallow(<Mole {...props} />)
  it("renders properly", () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe("when clicking on the `onMoleClick`", () => {
    it("should call the `onClick` function", () => {
      wrapper.simulate("click")

      expect(onMoleClick).toHaveBeenCalled()
    })
  })

  describe("when setting an `active` prop on a `Mole`", () => {
    it("should have a different `top` style property", () => {
      expect.assertions(3)

      expect(wrapper.find(".mole-image").exists()).toBe(true)
      expect(wrapper.find(".mole-image")).toHaveStyleRule("top", "80px")

      wrapper.setProps({active: true})

      expect(wrapper.find(".mole-image")).toHaveStyleRule("top", "0")
    })
  })
})
