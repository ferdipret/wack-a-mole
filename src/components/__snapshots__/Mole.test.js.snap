// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mole renders properly 1`] = `
<styled.div
  onClick={[Function]}
  success={false}
>
  <styled.div>
    <styled.div
      className="bomb-img"
    />
  </styled.div>
  <styled.div>
    <styled.div
      active={false}
      className="mole-image"
    />
  </styled.div>
  <styled.div />
</styled.div>
`;
